vortex-array/src/array/constant/canonical.rs

```diff
@@ -1,62 +1,100 @@
-                DType::Utf8(validity.nullability()),
-            )));
-        }
+        match self.scalar_value() {
+            ScalarValue::Bool(b) => Ok(Canonical::Bool(BoolArray::from_vec(
+                vec![*b; self.len()],
+                validity,
+            ))),
+            ScalarValue::Primitive(pvalue) => {
+                let ptype = if let DType::Primitive(ptype, _) = dtype {
+                    ptype
+                } else {
+                    vortex_bail!(
+                        "constant array with dtype {} but primitive value {}",
+                        dtype,
+                        pvalue
+                    );
+                };
 
-        if let Ok(b) = BinaryScalar::try_from(self.scalar()) {
-            let value = b.value();
-            let const_value = value.as_ref().map(|v| v.as_slice());
+                match_each_native_ptype!(ptype, |$P| {
+                    Ok(Canonical::Primitive(PrimitiveArray::from_vec::<$P>(
+                        vec![$P::try_from(*pvalue).unwrap_or_else(|_| $P::default()); self.len()],
+                        validity,
+                    )))
+                })
+            }
+            ScalarValue::Buffer(value) => {
+                let const_value = value.as_slice();
 
-            return Ok(Canonical::VarBin(VarBinArray::from_iter(
-                iter::repeat(const_value).take(self.len()),
-                DType::Binary(validity.nullability()),
-            )));
-        }
+                Ok(Canonical::VarBin(VarBinArray::from_iter_nonnull(
+                    iter::repeat(const_value).take(self.len()),
+                    dtype,
+                )))
+            }
+            ScalarValue::BufferString(value) => {
+                let const_value = value.as_bytes();
 
-        if let Ok(ptype) = PType::try_from(self.scalar().dtype()) {
-            return match_each_native_ptype!(ptype, |$P| {
-                Ok(Canonical::Primitive(PrimitiveArray::from_vec::<$P>(
-                    vec![$P::try_from(self.scalar()).unwrap_or_else(|_| $P::default()); self.len()],
-                    validity,
+                Ok(Canonical::VarBin(VarBinArray::from_iter_nonnull(
+                    iter::repeat(const_value).take(self.len()),
+                    dtype,
                 )))
-            });
-        }
+            }
+            ScalarValue::List(_) => vortex_bail!("Unsupported scalar type {}", dtype),
+            ScalarValue::Null => {
+                if !dtype.is_nullable() {
+                    vortex_bail!("dtype is non-nullable but value is null: {}", dtype)
+                }
 
-        vortex_bail!("Unsupported scalar type {}", self.dtype())
+                match dtype {
+                    DType::Null => Ok(Canonical::Null(NullArray::new(self.len()))),
+                    DType::Bool(_) => Ok(Canonical::Bool(BoolArray::from_vec(
+                        vec![true; self.len()],
```

Comment:
bool::default() is false

Reply:
@danking did you forget to push a change? There's still a diff here